### PR TITLE
Scope ReactInstanceManager to ReactActivity

### PR DESCRIPTION
**Pull request by cmc5788**
Android lifecycle for non-singleinstance or singletask Activity allows for some lifecycle overlap between same-Activity instances when a new Activity is immediately replacing an old instance, such as in the case of certain intents that can be used to launch an Activity (some popular third-party launchers such as NovaLauncher are easier to reproduce this with).

In these cases, the new Activity instance's create may be called before old instance's destroy, so it is NOT safe to use a singleton ReactInstanceManager scoped to the Application - if new Activity is created before old Activity is destroyed, very bad things will happen.

TLDR; Android lifecycle provides no guarantee that a dying Activity will have its onDestroy called before the new instance's onCreate is called in certain cases. ReactInstanceManager should be per-activity, and new Activity instances should have no potential to access to previous instance's ReactInstanceManager.

This PR fixes the immediate issue, but should probably consider doing away with the entire concept of casting Application to ReactApplication and getting dependencies that way - it's poor architecture, and dangerous with the Android lifecycle.

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @foghina and @andreicoman11 to be potential reviewers.

**Comment by facebook-github-bot-5** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by foghina** [comment]
If you're just launching a new `Activity` on top of the current one, the current one might not be destroyed at all, just paused. This is expected. However, **pausing** should definitely happen before the new activity is created/resumed. This is why we set the current activity to `null` in `onHostPause`.

>  very bad things will happen

Can you elaborate?

**Comment by facebook-github-bot-1** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by cmc5788** [comment]
@foghina Yes - in either case, `ReactNativeHost.clear()` being called in another Activity instance's `onDestroy` after the `onCreate` of a new Activity will cause any subsequent calls to `getReactNativeHost().getReactInstanceManager()` in the newest instance of the Activity to lazily initialize a new `ReactNativeInstanceManager` instance at some unspecified point in that new Activity's lifecycle. There will be overlapping Activity lifecycles calling through to the same instance of `ReactNativeInstanceManager`.

**Comment by foghina** [comment]
Argh, I see. We don't actually call `clear()` in `onDestroy()` internally. Perhaps that would be a better fix? Sharing a `ReactInstanceManager` through the host really should work, the approach in this PR is not good.

**Comment by cmc5788** [comment]
@foghina see my updated comment. Even if you don't clear in onDestroy, the new instance being created would hit that assert with a shared instance.

**Comment by foghina** [comment]
It wouldn't hit that assert, because `startReactApplication` checks `hasStartedCreatingInitialContext()` before calling `createReactContextInBackground()`.

Anyway, I get the point. Non-foreground activities can be destroyed at random times by the OS (e.g. under memory pressure), which will cause the foreground activity to lose its instance manager, or at least have it put into `BEFORE_CREATE` lifecycle state.

However, this is still not the correct way to fix this. The correct way would be to:

1. Not call `clear()` in `onDestroy()`.
2. Find a way to **not** call `onHostDestroy` if another client is using the `ReactInstanceManager` when an activity is destroyed. This can be done for example by adding a host tracking mechanism in `ReactInstanceManager` and ignoring lifecycle state changes from non-foreground activities. Not sure if this is the best approach though, you could also check in `ReactActivity#onDestroy` whether there is another user of the `ReactInstanceManager` and not call `onHostDestroy`.

At this point we're not going to merge this. If you feel like implementing the above, another PR would be welcome.

A workaround for you would be to implement your Activity from scratch and handle the `ReactInstanceManager` lifecycle however you want.

**Comment by cmc5788** [comment]
Okay, thanks. I'll work up another PR.
